In [1]:
%load_ext autoreload
%autoreload 1

import sys
sys.path.append("../utils/")

from sklearn.preprocessing import StandardScaler, Normalizer, MinMaxScaler, MaxAbsScaler, RobustScaler

from GraphAPI import GraphCreator
from graph_helpers import *
from evaluations import *

%aimport GraphAPI
%aimport graph_helpers
%aimport evaluations

## Generating Graph from Entry Point

1. We initialize our GraphCreator class and check how many new nodes we will need to query. 

In [2]:
# include_see_also=False used for validation below.
# in deployment, include_see_also should be set to True
gc = GraphCreator("Function (music)",
                  include_see_also=False, max_recursive_requests=50)
print("Number of Links to Search:", len(gc.next_links), "\n\n")
print(list(gc.primary_nodes.keys()), "\n\n")
print(gc.see_also_articles)

Number of Links to Search: 178 


['Chord (music)', 'Scale degree', 'Tonal centre'] 


['Common practice period', 'Constant structure', 'Diatonic and chromatic', 'Nondominant seventh chord', 'Secondary dominant', 'Subsidiary chord', 'Roman numeral analysis']


2. We query all the nodes linked to/from the entry point (expand our network one level for each node).

In [3]:
gc.expand_network_threaded(threads=20, chunk_size=1)
print("Number of Links After Expansion: ", len(gc.graph.nodes))

Number of Links After Expansion:  54655


3. Since some nodes will likely have linked to articles through a redirect link, we need to traverse our graph and ensure that all redirects are assigned to the correct nodes. Once all redirects have been dealt with, we remove any old redirect nodes. 

In [4]:
gc.redraw_redirects()

4. Edges are weighted by how many categories two connected nodes have in common. Once we have all our nodes, and we have dealt with redirects, we can add edge weights for our entire graph. 

In [5]:
gc.update_edge_weights()
gc.get_edge_weights().head()

source_node         target_node  edge_weight
0          Ostinato        Figured bass            5
1      Figured bass            Ostinato            5
2  Circle of fifths        Parallel key            4
3       Block chord       So What chord            4
4      Parallel key  Modulation (music)            4

# Getting Our Feature Set

There are two options when generating the feature set:

1. we can generate a standard feature set with only the features themselves. To do this, have the `rank` parameter set to `False`.
2. We can generate a ranked feature set (set `rank` equal to `True`). For each parameter, this will rank them in order of _best_ to _worst_ (this could be ascending or descending, depending on the context of the feature).

After running `get_features_df`, the feature set will be saved in the GraphCreator instance as `feature_df`

In [6]:
features_df = gc.get_features_df(rank=False)

# Similarity Rank

Two articles are more similar the more categories they share and the closer they are to each other. 

In [9]:
gc.rank_similarity()
gc.features_df.sort_values("similarity_rank", ascending=False)

node  degree  \
72                      Function (music)     240   
124                    Predominant chord     117   
79         Parallel and counter parallel     220   
117                           Supertonic     176   
115                             Subtonic     181   
51                       Secondary chord     298   
86                               Mediant     212   
142                          Terzschritt      49   
134                         Tonicization      74   
57                            Submediant     276   
52                          Leading-tone     298   
150                    Riemannian theory      40   
28                           Subdominant     396   
25                      Dominant (music)     495   
18                         Tonic (music)     615   
116                     Subsidiary chord     180   
110                 Mixed-interval chord     185   
113                      Complexe sonore     182   
109                          Dream chord     186   
112                             Secundal     184   
101                        Primary triad     192   
97                          Bridge chord     198   
96                  Common chord (music)     199   
98                        Approach chord     197   
106                Northern lights chord     187   
91                        Factor (chord)     204   
99             Nondominant seventh chord     193   
81                     Chromatic mediant     218   
107       Diminished major seventh chord     187   
105                      Seven six chord     187   
...                                  ...     ...   
28530    Institute for Historical Review       1   
28531                      Mixed economy       1   
28532                             Recife       1   
28533                   Please Please Me       1   
28534                       Eileen Chang       1   
28536           Dave Thompson (comedian)       1   
28537                      Magnetic flux       1   
28538                        Boyle's law       1   
28539                  Elastic collision       1   
28540                Inelastic collision       1   
28541                Neonatal heel prick       1   
28542                Equations of motion       1   
37754                    Function Drinks       1   
37746          Functionality (chemistry)       1   
37755               Function (sociology)       1   
37748                         Subroutine       1   
37747           Functor (disambiguation)       1   
37771             Function (engineering)       1   
37745     Functionalism (disambiguation)       1   
37744  Functional group (disambiguation)       1   
37743        Functional (disambiguation)       1   
37742                    Function object       1   
37741                     Function model       1   
37740                       Function key       1   
37739                      Function hall       1   
37738                    Function (song)       1   
37769             Boneyard Beach (album)       1   
37770                 Function (biology)       1   
37772                Function (language)       1   
37757                Function (musician)       1   

       category_matches_with_source  in_edges  out_edges  \
72                                1       116        124   
124                               1        16        101   
79                                1       100        120   
117                               1       139         37   
115                               1       128         53   
51                                1       115        183   
86                                1       167         45   
142                               1        21         28   
134                               1        38         36   
57                                1       153        123   
52                                1       147        151   
150                               1        13         27   
28                             

# Scaling Features

We can easily scale our each of our features through the `scale_features_df` method. It will default to `Standard Scaler`, but we can specify alternate scalers in the `scaler` parameter.  

In [8]:
scaled_feature_df = gc.scale_features_df(scaler=Normalizer, copy=True) # Makes a copy of the df
scaled_feature_df.sort_values("similarity_rank", ascending=False).reset_index().drop("index", axis=1)

node    degree  \
0                                    Riemannian theory  0.799644   
1                                          Terzschritt  0.813246   
2                                     Function (music)  0.816197   
3                                         Tonicization  0.816180   
4                                    Predominant chord  0.752861   
5                                           Supertonic  0.774263   
6                                             Subtonic  0.794049   
7                        Parallel and counter parallel  0.815342   
8                                              Mediant  0.774824   
9                                           Submediant  0.814877   
10                                    Percy Goetschius  0.795534   
11                                     Secondary chord  0.809486   
12                                             Ragtime  0.772997   
13                                       Maurice Ravel  0.801762   
14                                      Duke Ellington  0.777004   
15                                         Miles Davis  0.777004   
16                                     Frédéric Chopin  0.806045   
17                           The Well-Tempered Clavier  0.784442   
18                                Augmentation (music)  0.789328   
19     The New Grove Dictionary of Music and Musicians  0.755240   
20                                          Modal jazz  0.761980   
21                                    Musical keyboard  0.784442   
22                                    Progressive rock  0.749977   
23                                      Interval cycle  0.777462   
24                                         The Beatles  0.749973   
25                                   Anhemitonic scale  0.776131   
26                                 Vincent Persichetti  0.754805   
27                                        Leading-tone  0.816457   
28                             Wolfgang Amadeus Mozart  0.763456   
29                                           Accordion  0.761980   
...                                                ...       ...   
54429                         Functor (disambiguation)  0.138675   
54430                                       Subroutine  0.138675   
54431                                  Function Drinks  0.138675   
54432                             Function (sociology)  0.138675   
54433                              Function (musician)  0.138675   
54434                           Boneyard Beach (album)  0.138675   
54435                               Function (biology)  0.138675   
54436                           Function (engineering)  0.138675   
54437                              Function (language)  0.138675   
54438                                   Diatonic scale  0.812495   
54439                                  Just intonation  0.805717   
54440                                Heinrich Schenker  0.810960   
54441                               Modulation (music)  0.771958   
54442                                    Simon Sechter  0.812718   
54443                                         Arpeggio  0.786649   
54444                                         Tonality  0.781336   
54445                                     Figured bass  0.803522   
54446                        Consonance and dissonance  0.807304   
54447                                         Function  0.713024   
54448                                      Sonata form  0.796163   
54449                                         Ostinato  0.802949   
54450                             Jean-Philippe Rameau  0.809648   
54451                                Chord progression  0.769850   
54452                                          C major  0.759203   
54453                                Arnold Schoenberg  0.776313   
54454                                      Key (music)  0.738348   
54455                          Index of music articles  0.726274   
54456                                 

___
# Validation

Here, we _validate_ our results. For many articles, we already have some user defined links that are highly related to the present article. These are found in the **See Also** section of several Wikipedia articles (some pages do not have them). These are not ordered in importance in any meaningful way, and there are no rating scores.

The intuition in this validation is as follows: 

> _Given that we know some articles are highly related from user input, if the recommendations provided by this system are valid, we would expect to see those **See Also** links ranked relatively high on our list._ 

_Note: This validation is not meant as **confirmation** or **evaluation** of the results. It only provides us one way of telling if the results we are seeing are reasonably valid. It is important to note that we cannot compare these results across two different articles, as those would be two entirely different network structures, likely with different human labeled links._  


In [ ]:
evaluate_metrics(scaled_feature_df, 
                 on=["similarity_rank", "centrality", "adjusted_reciprocity", "page_rank", "shortest_path_length_from_entry", "jaccard_similarity"], 
                 targets=gc.see_also_articles)

The chart generated above compares different ranking metrics (left index) for a given article. The most important column, `score`, provides a fast way for us to compare these different metrics. 

For example, if we see a _score_ of 0.98 for a given ranking metric, The following statement would be true:

> All of the human labeled **See Also** links are present within the top 98% of our recommendations. 

Since the human labeled links comprise a range, it is not possible to get a score of 100%. The `max score possible` column indicates the score that would be achieved if all the human labeled _See Also_ links appeared at the top of our recommendations without any other links intervening.   

The `difference` column is an alternative way of looking at the score. If we had a 0.02 in this column, we could say:

> All the human labeled **See Also** links are contained within the top 2% of our recommendations. 

`Total targets` is the number of human labeled _See Also_ links. 

Because it is possible that different metrics could have similar scores, we want a way to break down the dispersion of the known related links to see if one metric does perform better than another. The trailing four columns provide us with a course way of measuring this dispersion. 

Each of these columns indicates the percentage of human labeled _See Also_ links captured within a given percentage of the top of our recommendations. For example, if we see a 0.92 in the `% targets in the top 1%` columns, we could say:

> 92% of the human labeled **See Also** links appear in the top 1% of our recommendations. 

The value of these columns is a follows - If two ranking metrics have similar scores, we _might_ consider the better performing one to be the one in which the majority of the human labeled links are higher in our recommendation list. 

___